File: SPOJ/ABA12C - Buying Apples.py
def f(cnt,kg,ar,n,k,dp):
    if cnt > n or kg > k:
        return 10000000
    if k==kg:
        return 0
    if dp[cnt][kg]!=10000000:
        return  dp[cnt][kg]
    else:
        for i in range(1,k+1):
            if ar[i]!=-1:
                dp[cnt][kg]=min(dp[cnt][kg],f(cnt+1,kg+i,ar,n,k,dp)+ar[i])
        return dp[cnt][kg]
 
 
t = int(input())
for test in range(t):
    n, k = map(int,input().split())
    cost = list(map(int,input().split()))
    dp = [ [10000000 for i in range(k+1)] for j in range(n+1)]
    ar = [0]*(k+1)
    for i in range(k):
        ar[i+1]=cost[i]
    res = f(0,0,ar,n,k,dp)
    if res >=10000000:
        print(-1)
    else:
        print(res) 

File: SPOJ/ABA12C - Buying Apples.cpp
    #include <bits/stdc++.h>
    #define ll long long
    using namespace std;
     
    ll dp[105][105],n,k,inf=100000000,ar[105];
     
    ll f(int cnt,int kg)
    {
        //cout<<kg<<" "<<k<<endl;
        if(cnt>n||kg>k)
        return inf;
        if(kg==k)
        return 0;
        //cout<<dp[cnt][kg]<<endl;
        if(dp[cnt][kg]!=inf)
        return dp[cnt][kg];
            for(int i=1;i<=k;i++)
            {
                if(ar[i]!=-1)
                dp[cnt][kg]=min(dp[cnt][kg],ar[i]+f(cnt+1,kg+i));
                //cout<<dp[cnt][kg]<<endl;
     
            }
     
            return dp[cnt][kg];
    }
     
    int main()
    {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        int t;
        cin>>t;
        while(t--)
        {
            cin>>n>>k;
            for(int i=1;i<=k;i++)
            cin>>ar[i];
            for(int i=0;i<=n;i++)
            {
                for(int j=0;j<=k;j++)
                dp[i][j]=inf;
            }
            ll res=f(0,0);
            if(res>=inf)
            cout<<"-1"<<endl;
            else
            cout<<res<<endl;
     
        }
     
        return 0;
    }
     
